pkg/tsdb/prometheus/prometheus.go

```diff
@@ -141,14 +141,36 @@ func (e *PrometheusExecutor) parseQuery(dsInfo *models.DataSource, query plugins
 			return nil, err
 		}
 
-		dsInterval, err := interval.GetIntervalFrom(dsInfo, queryModel.Model, time.Second*15)
+		stepMode := queryModel.Model.Get("stepMode").MustString("min")
+		intervalFactor := queryModel.Model.Get("intervalFactor").MustInt64(1)
+		hasQueryInterval := queryModel.Model.Get("interval").MustString("") != ""
+		var adjustedIntervalValue time.Duration
+
+		// Calculate interval value from query or data source settings or use default value
+		intervalValue, err := interval.GetIntervalFrom(dsInfo, queryModel.Model, time.Second*15)
 		if err != nil {
 			return nil, err
 		}
 
-		intervalFactor := queryModel.Model.Get("intervalFactor").MustInt64(1)
-		interval := e.intervalCalculator.Calculate(*query.TimeRange, dsInterval)
-		step := time.Duration(int64(interval.Value) * intervalFactor)
+		if hasQueryInterval && stepMode == "max" {
+			// With max stepMode - use interval calculator to get ideal interval based on time range
+			rangeIntervalValue := e.intervalCalculator.Calculate(*query.TimeRange, 0).Value
+			// If rangeIntervalValue is smaller than max interval value, use it
+			if rangeIntervalValue < intervalValue {
+				adjustedIntervalValue = rangeIntervalValue
+				// Otherwise use max intervalValue
+			} else {
+				adjustedIntervalValue = intervalValue
+			}
+		} else if hasQueryInterval && stepMode == "exact" {
+			// With exact stepMode - use passed interval value
+			adjustedIntervalValue = intervalValue
+		} else {
+			// With default stepMode - use interval calculator to get ideal interval based on time range and passed value
+			adjustedIntervalValue = e.intervalCalculator.Calculate(*query.TimeRange, intervalValue).Value
```

Comment:
This was a bit confusing to me. Basically we have branches that does max/exact and then one that does both min and "auto". I thought this was missing one `else if stepMode == "min"` branch. Would it make sense to just keep this inside `Calculate` function? So it would be `Calculate(timerange, interval, stepMode)` or something like that and all this logic would be there?

Reply:
@aocenas  the `Calculate` function is a shared function used by many datasources (which do not have the step-mode), so maybe create a "prometheusCalculate" function that wraps it and does it's thing?

Reply:
well I guess we can have some defaults there or something? Right now it has a default stepMode == min.

Reply:
It would need different signature but I assume this does not have to be backward compatible does it?

Reply:
As none of the other data sources support other interval type than `min`, and as far as I know there is no plan to support it, I tried to do minimal changes that would affect only `prometheus.go`. Moreover, we plan to switch prometheus to backend data source soon, which means that most of the code is going to change/improve/grow. I didn't want to do changes that would affect other data sources that might change again. So I looked at this as temporary smallest-impact solution that will be refactored with all of the incoming changes. 

@aocenas  Do you think it makes sense to update Calculate method now, or would you be okay with this solution that is going to be changed and cleaned up when we move prometheus to be backend data source?

Reply:
This also still need an update cause in the frontend PR we are always using safe interval, if calculated interval is unsafe. Therefore we need to implement this also here. Will do it after we decide on how to progress.

Reply:
I don't think this needs to affect other data sources. I mean yeah if the signature changes than there will be change but I don't think that is significant. Ie if have function that calculates interval that right now is specialized to compute interval from range and using also minInterval, generalizing it to function that calculates interval and takes different limit values "min/max/exact" still makes sense to me from stand point of how the code is structured and what the current abstraction is. Which data source uses which limits is sort of inconsequential, we have a function to calculate the interval and different data sources will use it in different way.

I mean this is not a strong difference/preference on my side but such code structure makes more sense to me, especially when looking at this particular set of `if else` where instead of what I expected min/max/exact/default cases it is max/exact/(min+default).

Reply:
one complication might be that it seems datasources are migrating to the "new api", and datasources on old-api use https://github.com/grafana/grafana/blob/main/pkg/tsdb/interval/interval.go#L57 and datasources on new-api use https://github.com/grafana/grafana/blob/main/pkg/tsdb/calculator.go#L56

Reply:
i do agree that the code with the IFs is hard to follow though :-(

Reply:
I have refactored it here: https://github.com/grafana/grafana/pull/36796/commits/045db09063fd26153b530b9b3f8ce825097d5c12. It definitely seems more readable. Also, prometheus currently use old api, so I refactored Calculate method in old api in `interval.go`. I merge main and fix merge conflicts for the ddata sources that use new api from `calculator.go`. Can you check it out.

Reply:
I have also added safe interval check as it is on frontend.